vector/src/main/java/org/apache/arrow/vector/compare/RangeEqualsVisitor.java

```diff
@@ -270,35 +271,24 @@ protected boolean compareRunEndEncodedVectors(Range range) {
     RunEndEncodedVector leftVector = (RunEndEncodedVector) left;
     RunEndEncodedVector rightVector = (RunEndEncodedVector) right;
 
-    final int leftRangeEnd = range.getLeftStart() + range.getLength();
-    final int rightRangeEnd = range.getRightStart() + range.getLength();
+    final RunEndEncodedVector.RangeIterator leftIterator =
+        new RangeIterator(leftVector, range.getLeftStart(), range.getLength());
+    final RunEndEncodedVector.RangeIterator rightIterator =
+        new RangeIterator(rightVector, range.getRightStart(), range.getLength());
 
     FieldVector leftValuesVector = leftVector.getValuesVector();
     FieldVector rightValuesVector = rightVector.getValuesVector();
 
     RangeEqualsVisitor innerVisitor = createInnerVisitor(leftValuesVector, rightValuesVector, null);
 
-    int leftLogicalIndex = range.getLeftStart();
-    int rightLogicalIndex = range.getRightStart();
+    while (leftIterator.nextRun() | rightIterator.nextRun()) {
```

Comment:
```suggestion
    while (leftIterator.nextRun() || rightIterator.nextRun()) {
```

Reply:
I suppose we already checked the vector lengths to be equal, and since we're requiring run lengths to be equal, we don't need to check for any leftover runs after the loop?

Reply:
@lidavidm The vector length is checked in `RangeEqualsVisitor.rangeEqual`. If we call vector.visit(RangeEqualsVisitor, Range)` directly, the argument checks are skipped. Using rangeEqual is the best practice, but we seem to be unable to limit users from using another method. Therefore, I'm also considering whether we need to add more boundary checks in Iterator. What do you think?

Reply:
I think it doesn't hurt to make sure that both iterators are at the end at the end of the loop as a sanity check.

Reply:
@lidavidm According to your suggestion, I should change it to `while (leftIterator.nextRun() && rightIterator.nextRun())` instead of `while (leftIterator.nextRun() || rightIterator.nextRun())`. The short-circuit feature of '||' will make only leftIterator move forward, while right stays still.